openlane/flows/synth_explore.py

```diff
@@ -0,0 +1,143 @@
+    Steps = [
+        Yosys.Synthesis,
+        OpenROAD.CheckSDCFiles,
+        OpenROAD.STAPrePNR,
+    ]
+
+    def run(
+        self,
+        initial_state: State,
+        **kwargs,
+    ) -> Tuple[State, List[Step]]:
+        step_list: List[Step] = []
+
+        self.progress_bar.set_max_stage_count(1)
+
+        synth_futures: List[Tuple[Config, Future[State]]] = []
+        self.progress_bar.start_stage("Synthesis Exploration")
+
+        options.set_condensed_mode(True)
+
+        for strategy in [
+            "AREA 0",
+            "AREA 1",
+            "AREA 2",
+            "AREA 3",
+            "DELAY 0",
+            "DELAY 1",
+            "DELAY 2",
+            "DELAY 3",
+            "DELAY 4",
+        ]:
+            config = self.config.copy(SYNTH_STRATEGY=strategy)
+
+            synth_step = Yosys.Synthesis(
+                config,
+                id=f"synthesis-{strategy}",
+                state_in=initial_state,
+            )
+            synth_future = self.start_step_async(synth_step)
+            step_list.append(synth_step)
+
+            sdc_step = OpenROAD.CheckSDCFiles(
+                config,
+                id=f"sdc-{strategy}",
+                state_in=synth_future,
+            )
+            sdc_future = self.start_step_async(sdc_step)
+            step_list.append(sdc_step)
+
+            sta_step = OpenROAD.STAPrePNR(
+                config,
+                state_in=sdc_future,
+                id=f"sta-{strategy}",
+            )
+
+            step_list.append(sta_step)
+            sta_future = self.start_step_async(sta_step)
+
+            synth_futures.append((config, sta_future))
+
+        results: Dict[str, Optional[Tuple[Decimal, Decimal, Decimal, Decimal]]] = {}
+        for config, future in synth_futures:
+            strategy = config["SYNTH_STRATEGY"]
+            results[strategy] = None
+            try:
+                state = future.result()
+                results[strategy] = (
+                    state.metrics["design__instance__count"],
+                    state.metrics["design__instance__area"],
+                    state.metrics["timing__setup__ws"],
+                    state.metrics["timing__setup__tns"],
+                )
+            except StepError:
+                pass  # None == failure
+        self.progress_bar.end_stage()
+        options.set_condensed_mode(False)
+
+        successful_results = {k: v for k, v in results.items() if v is not None}
+        min_gates = min(map(lambda x: x[0], successful_results.values()))
+        min_area = min(map(lambda x: x[1], successful_results.values()))
+        max_slack = max(map(lambda x: x[2], successful_results.values()))
+        max_tns = max(map(lambda x: x[3], successful_results.values()))
+
+        table = rich.table.Table()
+        table.add_column("SYNTH_STRATEGY")
+        table.add_column("Gates")
+        table.add_column("Area (µm²)")
+        table.add_column("Setup Slack (ns)")
```

Comment:
Also include hold?

Reply:
Discussed with Passant and Hosni- Hold is kind of useless at this stage